drivers/spi/spi_pl022.c

```diff
@@ -0,0 +1,334 @@
+	sys_write32(temp, address);
+}
+
+static void rx_buffer_flush(u32_t base)
+{
+	u32_t stat_reg = SPI_REG_ADDR(base, SPI_SR_OFFSET);
+
+	/* Read data until Rx FIFO is Empty */
+	while (sys_test_bit(stat_reg, RX_NOEMPTY))
+		sys_read32(SPI_REG_ADDR(base, SPI_DR_OFFSET));
+}
+
+static void wait_for_sync(u32_t base)
+{
+	u32_t stat_reg = SPI_REG_ADDR(base, SPI_SR_OFFSET);
+
+	/* wait until busy flag is clear */
+	while (sys_test_bit(stat_reg, BUSY))
+		;
+}
+
+static bool spi_transfer_complete(struct device *dev)
+{
+	struct spi_pl022_data *data = dev->driver_data;
+
+	if (spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx)) {
+		return 1;
+	}
+	return 0;
+}
+
+static void pull_data(struct device *dev)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	struct spi_pl022_data *data = dev->driver_data;
+	u32_t base = cfg->base;
+	u16_t value;
+
+	/* Check Rx FIFO is Not Empty */
+	while (sys_test_bit(SPI_REG_ADDR(base, SPI_SR_OFFSET), RX_NOEMPTY)) {
+
+		value = sys_read16(SPI_REG_ADDR(base, SPI_DR_OFFSET));
+
+		if (spi_context_rx_buf_on(&data->ctx)) {
+			UNALIGNED_PUT(value, (u8_t *)data->ctx.rx_buf);
+		}
+		spi_context_update_rx(&data->ctx, 1, 1);
+	}
+}
+
+static void push_data(struct device *dev)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	struct spi_pl022_data *data = dev->driver_data;
+	u32_t base = cfg->base;
+	u16_t value;
+
+	/* Check Tx FIFO is Not Full */
+	if (sys_test_bit(SPI_REG_ADDR(base, SPI_SR_OFFSET), TX_NOFULL)) {
+		value = 0;
+
+		/* Hold, if Rx FIFO Full */
+		if (sys_test_bit(SPI_REG_ADDR(base, SPI_SR_OFFSET), RX_FULL)) {
+			return;
+		}
+
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			value = UNALIGNED_GET((u8_t *)(data->ctx.tx_buf));
+		}
+
+		sys_write16(value, SPI_REG_ADDR(base, SPI_DR_OFFSET));
+		spi_context_update_tx(&data->ctx, 1, 1);
+	}
+}
+
+static void set_frequency(struct device *dev, u32_t freq)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	u32_t base = cfg->base;
+	u32_t i, j;
+	u32_t spi_clock;
+
+	/* Get periperal clock frequency */
+	clock_control_get_rate(device_get_binding(DT_CLOCK_CONTROL_LABEL),
+			       (clock_control_subsys_t) &cfg->pclk, &spi_clock);
+
+	for (i = CLK_PRESCALE_MIN; i <= CLK_PRESCALE_MAX; i += 2)
+		for (j = SERIAL_CLKRATE_MIN; j <= SERIAL_CLKRATE_MAX; j++)
+			if (freq >= (spi_clock / (i * j))) {
+				/* Set optimal div value */
+				sys_set_bits(
+					SPI_REG_ADDR(base, SPI_CPSR_OFFSET),
+					BYTE_MASK, CLK_PRESCALE_SHIFT, i);
+
+				/* Set optimal SCR value */
+				sys_set_bits(
+					SPI_REG_ADDR(base, SPI_CR0_OFFSET),
+					BYTE_MASK, SERIAL_CLKRATE_SHIFT, j - 1);
+				return;
+			}
+	/* Set div value to default if frequency is less*/
+	sys_set_bits(SPI_REG_ADDR(base, SPI_CPSR_OFFSET), 0xFF,
+		     CLK_PRESCALE_SHIFT, CLK_PRESCALE_MAX);
+
+	/* Set SCR value to default if frequency is less */
+	sys_set_bits(SPI_REG_ADDR(base, SPI_CR0_OFFSET), 0xFF,
+		     SERIAL_CLKRATE_SHIFT, SERIAL_CLKRATE_MAX);
```

Comment:
and use clk_rate also here (probably with a test as well here, since it seems it should not remove 1 if it's max)

Reply:
fixed